gplog/gplog.go

```diff
@@ -0,0 +1,270 @@
+	 */
+	errorCode = 0
+	// Singleton logger used by any package or utility that calls InitializeLogging
+	logger *GpLogger
+)
+
+const (
+	/*
+	 * The following constants representing the current logging level, and are
+	 * cumulative (that is, setting the log level to Debug will print all Error-,
+	 * Info-, and Verbose-level messages in addition to Debug-level messages).
+	 */
+	LOGERROR = iota
+	LOGINFO
+	LOGVERBOSE
+	LOGDEBUG
+)
+
+/*
+ * Leveled logging output functions using the above log levels are implemented
+ * below.  Info(), Verbose(), and Debug() print messages when the log level is
+ * set at or above the log level matching their names.  Warn(), Error(), and
+ * Fatal() always print their messages regardless of the current log level.
+ *
+ * The intended usage of these functions is as follows:
+ * - Info: Messages that should always be written unless the user explicitly
+ *         suppresses output
+ * - Verbose: More detailed messages that are mostly useful to the user, e.g.
+ *            printing information about a function's substeps for progress tracking.
+ * - Debug: More detailed messages that are mostly useful to developers, e.g.
+ *          noting that a function has been called with certain arguments.
+ * - Warn: Messages indicating unusual but not incorrect behavior that a user
+ *         may want to know, e.g. that certain steps are skipped when using
+ *         certain flags.  These messages are shown even if output is suppressed.
+ * - Error: Messages indicating that an error has occurred, but that the program
+ *          can continue, e.g. one table failed to back up but others succeeded.
+ * - Fatal: Messages indicating that the program cannot proceed, e.g. the database
+ *          cannot be reached.  This function will exit the program after printing
+ *          the error message.
+ */
+
+type GpLogger struct {
+	logStdout   *log.Logger
+	logStderr   *log.Logger
+	logFile     *log.Logger
+	logFileName string
+	verbosity   int
+	header      string
+}
+
+/*
+ * Logger initialization/helper functions
+ */
+
+/*
+ * Multiple calls to InitializeLogging can be made if desired; the first call
+ * will initialize the logger as a singleton and subsequent calls will return
+ * the same Logger instance.
+ */
+func InitializeLogging(program string, logdir string) {
+	if logger != nil {
+		return
+	}
+	currentUser, _ := operating.System.CurrentUser()
+	if logdir == "" {
+		logdir = fmt.Sprintf("%s/gpAdminLogs", currentUser.HomeDir)
+	}
+
+	createLogDirectory(logdir)
+	timestamp := operating.System.Now().Format("20060102")
+	logfile := fmt.Sprintf("%s/%s_%s.log", logdir, program, timestamp)
+	logFileHandle := openLogFile(logfile)
+
+	logger = NewLogger(os.Stdout, os.Stderr, logFileHandle, logfile, LOGINFO, program)
+}
+
+func SetLogger(log *GpLogger) {
+	logger = log
+}
+
+// This function should only be used for testing purposes
+func GetLogger() *GpLogger {
```

Comment:
I would somewhat rather just take this out. I'm not sure it provides a lot of value.